### DISCUSSION TITLE
How to globally set Attributes for ALL Spans ?

### DISCUSSION THREAD

**Discussion by bmistry13** [open-telemetry/opentelemetry-java-instrumentation#2163] [Q&A]
I have a use case where I would like to tag all spans with user session (session_uuid).  This will act as TAG for all span. When I search in zipkin, by span attributes, I will search for all spans that have session-id == 'some_session_id'.  

Is there a thread-local that can be set per-thread for the span that will automatically set attributes?

**Comment by iNikem** [comment]
One option is to provide your own [`SpanProcessor`](https://github.com/open-telemetry/opentelemetry-java/blob/main/sdk/trace/src/main/java/io/opentelemetry/sdk/trace/SpanProcessor.java#L59).

**Reply by anuraaga** [reply]
@iNikem Can an agent user do it (reasonably)?

**Reply by iNikem** [reply]
Any solution will be programmatic. The only barrier is to add new `SpanProcessor` to the existing SDK. I am not yet ready to say how easy it is, but IMO it is certainly doable.

**Reply by HaloFour** [reply]
It's a shame [SimpleSpanProcessor](https://github.com/open-telemetry/opentelemetry-java/blob/main/sdk/trace/src/main/java/io/opentelemetry/sdk/trace/export/SimpleSpanProcessor.java) is final, otherwise it would be trivial to derive from it and override `isStartRequired` and `onStart` to populate some default attributes.  Still relatively trivial to write an implementation of `SpanProcessor` that would delegate to `SimpleSpanProcessor`.

**Reply by iNikem** [reply]
You don't need to delegate. Just implement your own `SpanProcessor` and do whatever you like it. `onStart` methods in `SpanProcessor` are totally independent on each other.

**Reply by HaloFour** [reply]
That's true, you could write a simpler implementation that returns `false` for `isEndRequired` and add that to a `MultiSpanProcessor` as well.  At worse you have to stub out some of the required methods you won't use, but a good chunk of them are defaulted.

**Reply by bmistry13** [reply]
@iNikem,  I was looking to use **SpanProcessor**, but do I need to export it I was looking at **SimpleSpanProcessor** which has an exporter to send span.  Please let me know.  is the following sufficient?


```
 new SpanProcessor(){
            @Override
            public void onStart(Context parentContext, ReadWriteSpan span) {
                MyContext ctx = MyCtxHolder.getCtx();
                if(ctx != null && StringUtils.isNotBlank(ctx.getUiSession())){
                    span.setAttribute("ui_session",ctx.getUiSession());
                }
            }

            @Override
            public boolean isStartRequired() {
                return true;
            }

            @Override
            public void onEnd(ReadableSpan span) {
            }

            @Override
            public boolean isEndRequired() {
                return true;
            }
        }
```

**Reply by iNikem** [reply]
Change `isEndRequired` to return false and everything is good :)

**Comment by HaloFour** [comment]
Would all of these spans belong to one parent span where you could set the attributes?  If so, sounds like it could fall under this request to add inheritable attributes to spans: https://github.com/open-telemetry/opentelemetry-specification/issues/1337

**Reply by bmistry13** [reply]
This is the correct span attribute (session_uuid) that should be inheritable across all children's spans.  It is like tagging all spans with session_uuid so can find all spans with session_uuid for a user session. That is the goal.

**Comment by bmistry13** [comment]
Hi All,

I have spring-slueth and open telemetry library, so my question is if I register SpanProcessor,  will both Java Agent and Spring-sleuth (which depends on Open telemetry lib), will both will have setup attribute.   I am a newbie to open telemetry  so I have few questions:

1) Does register one SpanProcessor will all Tracers will use it ?

[Example](https://github.com/open-telemetry/opentelemetry-java/blob/68ee47108a808ad22bfa19618107263ac173e72f/examples/sdk-usage/src/main/java/io/opentelemetry/sdk/example/ConfigureSpanProcessorExample.java )  

2) From the above example, I would have the first register in my main method in the Spring-boot application before any tracer is created?

Again, I am new to Open telemetry and trying to understand SpanProcessor.
